### PR TITLE
Add Close() method to gogitRepository

### LINKED ISSUE #7947: Transfer the ownership  转移仓库所有权 do not work in V1.90~V1.92
i loopback to v1.83,it works.

### PR DESCRIPTION
**Pull request by zeripath**
In investigating #7947 it has become clear that the storage component of go-git repositories needs closing. 

This PR adds this Close function and adds the Close functions as necessary.

In TransferOwnership the ctx.Repo.GitRepo is closed if it is open to help prevent the risk of multiple open files.

Fixes #7947

### CODE REVIEW

File: models/repo_tag.go [hunk 1]
```diff
@@ -14,6 +14,7 @@ func GetTagsByPath(path string) ([]*git.Tag, error) {
 	if err != nil {
 		return nil, err
 	}
+	defer gitRepo.Close()
 
 	return gitRepo.GetTagInfos()
```
Comment (guillep2k) [inline models/repo_tag.go]: This could be a problem since Tag contains a reference to Repository.
```
type Tag struct {
	Name    string
	ID      SHA1
	repo    *Repository
	Object  SHA1 // The id of this commit object
	Type    string
	Tagger  *Signature
	Message string
}
```

Comment (zeripath) [reply models/repo_tag.go]: Probably is.

I'll take another look.

File: routers/repo/setting.go [hunk 2]
```diff
@@ -378,6 +378,11 @@ func SettingsPost(ctx *context.Context, form auth.RepoSettingForm) {
 		}
 
 		oldOwnerID := ctx.Repo.Owner.ID
+		// Close the GitRepo if ope
```
Comment (guillep2k) [inline routers/repo/setting.go]:
```suggestion
		// Close the GitRepo if open
```

File: routers/repo/setting.go [hunk 3]
```diff
@@ -378,6 +378,11 @@ func SettingsPost(ctx *context.Context, form auth.RepoSettingForm) {
...
+		if ctx.Repo.GitRepo != nil {
+			ctx.Repo.GitRepo.Close()
```
Comment (guillep2k) [inline routers/repo/setting.go:389]: I think this is necessary in `SettingsPost()` for the rename action as well.

Comment (lunny) [reply routers/repo/setting.go:389]: Should we close the git repository on the router? Or maybe we should close it on middleware which load it after invoke the router.

File: modules/context/repo.go [hunk 4]
```diff
@@ -426,10 +426,16 @@ func RepoAssignment() macaron.Handler {
 			return
 		}
 		ctx.Repo.GitRepo = gitRepo
+		defer func() {
+			if ctx.Repo.GitRepo == nil {
+				ctx.Repo.GitRepo.Close()
```
Comment (filipnavara) [inline modules/context/repo.go:461]: `!= nil` ?

File: routers/repo/setting.go (hunk 3 above)
Comment (zeripath) [reply routers/repo/setting.go:389]: @guillep2k done now.

@lunny I thought I had done this but I realise that it must have not saved. See the changes in modules/context/repo.go

File: modules/context/repo.go (hunk 4 above)
Comment (zeripath) [reply modules/context/repo.go:461]: oops.

Comment (zeripath) [reply modules/context/repo.go:461]: It would also help if I actually ran the defer function...

**Comment by Sebazzz** [comment]
As you requested in #8915, I'm taking a look 👍

Four remarks:
- Any reason for littering the close calls throughout the source code? This makes it quite error prone for any existing or new development and also makes merging this fix between branches harder. I did not know you opened this pull request at the time, but in #8915 I took the approach of unifying this clean-up in a single middleware. You might want to consider that approach. 
- I think that only for renaming and closing the repository needs to be closed prematurely. Else it is fine. 
- Closing the storage should be pretty harmless, go-git re-opens the necessary pack files automatically if necessary I think.
- Same as #8915, this suffers from the potential race condition where a rename or transfer could conflict which currently running page views. I don't think there is a good simple solution (beside locking) but it is also a minor issue in my opinion because renames and transfers are quite rare and can easily be re-tried. Perhaps we could handle the error better instead of falling back on 500 like "Access denied. Please try again and if that does not work contact your system administrator.".

One question:
Were you able to test this on Windows?

File: models/repo_tag.go (hunk 1 above)
Comment (zeripath) [reply models/repo_tag.go]: Oh dear GetTagInfos is a terrible implementation. It needs rewriting...

**Comment by zeripath** [comment]
@Sebazzz There is is more than one place that a gogitRepo is opened. That's the reason for all the closes. ctx.Repo.GitRepo is not the only use a gogitrepo - in fact it's more than probable that we open and reopen the same gogitrepo. (Which may lead to locks - I need to check in general)

If you open a resource you are supposed to close it - It's not premature closing to Close a resource when you're done with it. The Garbage Collector is not supposed to do the closing for you as you cannot be sure when that is going to run.

**Comment by lunny** [comment]
@zeripath This should also be back port to v1.9

File: modules/git/blame.go [hunk 5]
```diff
@@ -87,10 +87,11 @@ func (r *BlameReader) Close() error {
 
 // CreateBlameReader creates reader for given repository, commit and file
 func CreateBlameReader(repoPath, commitID, file string) (*BlameReader, error) {
-	_, err := OpenRepository(repoPath)
+	gitRepo, err := OpenRepository(repoPath)
```
Comment (lunny) [inline modules/git/blame.go:90]: Just check if the repoPath exist?

File: services/mirror/mirror.go [hunk 6]
```diff
@@ -290,97 +292,103 @@ func Update() {
 func SyncMirrors() {
 	// Start listening on new sync requests.
 	for repoID := range mirrorQueue.Queue() {
-		log.Trace("SyncMirrors [repo_id: %v]", repoID)
-		mirrorQueue.Remove(repoID)
+		syncMirror(repoID)
```
Comment (lunny) [inline services/mirror/mirror.go:295]: I prefer to send another PR to improve this code.

Comment (zeripath) [reply services/mirror/mirror.go:295]: It's not a real refactor - extracting the function simply makes closing the git repo easier - otherwise you can't just use defer.

File: modules/git/repo.go [hunk 7]
```diff
@@ -125,7 +125,7 @@ func OpenRepository(repoPath string) (*Repository, error) {
 
 // Close this repository, in particular close the underlying gogitStorage if this is not nil
 func (repo *Repository) Close() {
-	if repo.gogitStorage == nil {
+	if repo == nil || repo.gogitStorage == nil {
```
Comment (guillep2k) [inline modules/git/repo.go:128]: This is weird. Can an object method in golang be called even when the object is nil? (e.g. in C++ you can't). I mean, is it possible for repo to be `nil` here?

**Review by guillep2k — APPROVED** [review]
Just one question. Otherwise LGTM.

**Comment by guillep2k** [comment]
> 
> 
> Just one question. Otherwise LGTM.

Oops! I forgot about `GetTagInfos()`...

File: modules/git/repo.go (hunk 7 above)
Comment (zeripath) [reply modules/git/repo.go:128]: It appears possible - The reason why I had to add it is because I had put `==` when I meant `!=` in an earlier iteration however, it genuinely appears to be the case:

https://play.golang.org/p/nCrm8x910Wt

I kinda feel it's ok to leave it in there.

File: modules/git/blame.go (hunk 5 above)
Comment (zeripath) [reply modules/git/blame.go:90]: Well I suspect OpenRepository does a bit more than just check if the directory exists - however I agree this is silly.

I will have a look.

**Comment by zeripath** [comment]
@guillep2k I hadn't forgotten and I've decided that function has to die.

I'm changing the one place that calls it to call the function directly from GitRepo.

File: routers/api/v1/repo/tag.go [hunk 8]
```diff
@@ -33,7 +33,7 @@ func ListTags(ctx *context.APIContext) {
 	// responses:
 	//   "200":
 	//     "$ref": "#/responses/TagList"
-	tags, err := ctx.Repo.Repository.GetTags()
+	tags, err := ctx.Repo.GitRepo.GetTagInfos()
```
Comment (guillep2k) [inline routers/api/v1/repo/tag.go:36]: So each `tags[]` element will contain a reference to the same `Repository` object, and it should be the same as `ctx`'s, so hopefully the `ctx.Repository.Close()` at the end/defer of the ctx should be enough to ensure that all files are closed. 🤔

Comment (zeripath) [reply routers/api/v1/repo/tag.go:36]: Yes. Any packfiles are kept open in the gogitStorage within the repo. Nowhere else as far as I can see.

**Comment by Sebazzz** [comment]
> 
> 
> @Sebazzz There is is more than one place that a gogitRepo is opened. That's the reason for all the closes. 

Yes, in the background jobs you are probably referring to. Don't those run in their own context anyway? I suppose there is some place that runs when a job is finished, so that would be a good place to add a call to close the repo.

 
> If you open a resource you are supposed to close it - It's not premature closing to Close a resource when you're done with it. The Garbage Collector is not supposed to do the closing for you as you cannot be sure when that is going to run.

I think you misunderstood 😉  I said "I think that only for renaming and closing the repository needs to be closed prematurely. Else it is fine." -> Only when renaming and transferring the repository the gitRepo needs to be closed earlier then the end of the request.

**Comment by lunny** [comment]
@Sebazzz Not only HTTP router context reference `git.OpenRepository`, cron tasks, pushes,global pullrequests. I'm afraid it's not enough to only close the context repository.

File: integrations/repofiles_delete_test.go [hunk 9]
```diff
@@ -111,6 +112,8 @@ func testDeleteRepoFileWithoutBranchNames(t *testing.T, u *url.URL) {
 	test.LoadRepoCommit(t, ctx)
 	test.LoadUser(t, ctx, 2)
 	test.LoadGitRepo(t, ctx)
+	defer ctx.Repo.GitRepo.Close()
```
Comment (lunny) [inline integrations/repofiles_delete_test.go:115]: How about add `MockContext.Close` and close repository there?

File: routers/api/v1/repo/tag.go (hunk 8 above)
Comment (lunny) [reply routers/api/v1/repo/tag.go:36]: Why this should be changed here?

Comment (zeripath) [reply routers/api/v1/repo/tag.go:36]: Because otherwise it's impossible to cleanly close the gogit.Repository.

File: modules/git/blame.go (hunk 5 above)
Comment (zeripath) [reply modules/git/blame.go:90]: TBH blame needs a total refactor. We never actually run this code not on a repository so this is somewhat redundant. Rather than refactor that here I think just ensuring that this is closed should be sufficient.

File: integrations/repofiles_delete_test.go (hunk 9 above)
Comment (zeripath) [reply integrations/repofiles_delete_test.go:115]: We would need to add a Close() to context.Context as MockContext actually produces a real context.

**Comment by guillep2k** [comment]
@Sebazzz It's true that this is fallible, but IMHO it's an improvement and _should_ fix #7947 in most scenarios. Is it possible for you to test this PR and try to reproduce the problem?

**Comment by Sebazzz** [comment]
Works well on Windows.

Operations performed:
- Move
- Transfer

Tested on the reproducible testcase from #7947.

**Comment by lunny** [comment]
Please resolve conflicts

**Comment by zeripath** [comment]
Resolved

**Review by sapk — APPROVED** [review]

**Comment by lunny** [comment]
Please send back port to v1.10 and v1.9
